modules/pref/src/main/NotificationPref.scala

```diff
@@ -0,0 +1,114 @@
+    "tournamentSoon",
+    "gameEvent",
+    "invitedStudy"
+  )
+
+  def fromForm(bell: Boolean, push: Boolean): Allows =
+    Allows((bell ?? BELL) | (push ?? PUSH))
+
+  def toForm(allows: Allows): Some[(Boolean, Boolean)] =
+    Some((allows.bell, allows.push))
+
+  def fromCode(code: Int) = Allows(code)
+
+case class NotifyAllows(userId: UserId, allows: Allows):
+  export allows.*
+
+// take care with NotificationPref field names - they map directly to db and ws channels
+
+case class NotificationPref(
+    privateMessage: Allows,
+    challenge: Allows,
+    mention: Allows,
+    streamStart: Allows,
+    tournamentSoon: Allows,
+    gameEvent: Allows,
+    invitedStudy: Allows,
+    correspondenceEmail: Int
+):
+
+  def allows(event: Event): Allows = allows(event.toString)
+
+  def allows(tpe: String): Allows =
+    val (key, values) = (this.productElementNames, this.productIterator)
+    while (key.hasNext) {
+      val value = values.next()
+      if (key.next() == tpe) return value.asInstanceOf[Allows]
+    }
+    Allows(0)
+
+object NotificationPref:
+  val BELL   = 1
+  val WEB    = 2
+  val DEVICE = 4
+  val PUSH   = WEB | DEVICE
+
+  sealed trait Event:
+    override def toString: String = // for matching db fields, channels
+      val typeName = getClass.getSimpleName
+      s"${typeName.charAt(0).toLower}${typeName.substring(1, typeName.length - 1)}" // strip $
+
+  case object PrivateMessage extends Event
+  case object Challenge      extends Event
+  case object Mention        extends Event
+  case object InvitedStudy   extends Event
+  case object StreamStart    extends Event
+  case object TournamentSoon extends Event
+  case object GameEvent      extends Event
+
+  lazy val default: NotificationPref = NotificationPref(
+    privateMessage = Allows(BELL | PUSH),
+    challenge = Allows(BELL | PUSH),
+    invitedStudy = Allows(BELL | PUSH),
+    mention = Allows(BELL | PUSH),
+    streamStart = Allows(BELL | PUSH),
+    tournamentSoon = Allows(PUSH),
+    gameEvent = Allows(PUSH),
+    correspondenceEmail = 0
+  )
+
+  // implicit private val AllowsBSONHandler =
+  //  lila.db.dsl.intAnyValHandler[Allows](_.value, Allows.apply)
+  given BSONHandler[Allows] =
+    lila.db.dsl.intAnyValHandler[Allows](_.value, Allows.apply)
+
+  given BSONDocumentHandler[NotificationPref] = Macros.handler
+
+  given notificationDataJsonWriter: OWrites[NotificationPref] =
+    OWrites[NotificationPref] { data =>
+      Json.obj(
+        "privateMessage"      -> allowsToJson(data.privateMessage),
+        "mention"             -> allowsToJson(data.mention),
+        "streamStart"         -> allowsToJson(data.streamStart),
+        "challenge"           -> allowsToJson(data.challenge),
+        "tournamentSoon"      -> allowsToJson(data.tournamentSoon),
+        "gameEvent"           -> allowsToJson(data.gameEvent),
+        "invitedStudy"        -> allowsToJson(data.invitedStudy),
+        "correspondenceEmail" -> (data.correspondenceEmail != 0)
```

Comment:
once correspondenceEmail is a Boolean, we'll be able to remove this function and replace it with `Json.writes[NotificationPref]`